include/lsst/afw/typehandling/HeteroMap.h

```diff
@@ -0,0 +1,700 @@
+ * "contains" call requesting an integer labeled "value", for example, will report no such integer if instead
+ * there is a string labeled "value". For Python compatibility, a HeteroMap does not store type information
+ * internally, instead relying on RTTI for type checking.
+ *
+ * All subclasses **must** guarantee, as a class invariant, that every value in the map is implicitly
+ * nothrow-convertible to the type indicated by its key. For example, MutableHeteroMap ensures this by
+ * appropriately templating all operations that create new key-value pairs.
+ *
+ * A HeteroMap may contain primitive types, strings, Storable, and shared pointers to Storable as
+ * values. It does not support unique pointers to Storable because such pointers are read destructively. For
+ * safety reasons, it may not contain references, C-style pointers, or arrays to any type. Due to
+ * implementation restrictions, `const` types (particularly pointers to `const` Storable) are not
+ * currently supported.
+ */
+// TODO: const keys should be possible in C++17 with std::variant
+template <typename K>
+class HeteroMap {
+public:
+    using key_type = K;
+    using size_type = std::size_t;
+    using difference_type = std::ptrdiff_t;
+
+    virtual ~HeteroMap() = default;
+
+    /**
+     * Return a reference to the mapped value of the element with key equal to `key`.
+     *
+     * @tparam T the type of the element mapped to `key`
+     * @param key the key of the element to find
+     *
+     * @return a reference to the `T` mapped to `key`, if one exists
+     *
+     * @throws pex::exceptions::OutOfRangeError Thrown if the map does not
+     *         have a `T` with the specified key
+     * @exceptsafe Provides strong exception safety.
+     *
+     * @note This implementation calls @ref unsafeLookup once, then uses templates
+     *       and RTTI to determine if the value is of the expected type.
+     *
+     * @{
+     */
+    template <typename T, typename std::enable_if_t<!IS_SMART_PTR<T>, int> = 0>
+    T& at(Key<K, T> const& key) {
+        // Both casts are safe; see Effective C++, Item 3
+        return const_cast<T&>(static_cast<const HeteroMap&>(*this).at(key));
+    }
+
+    // Can't partially specialize method templates, rely on enable_if to avoid duplicates
+    template <typename T,
+              typename std::enable_if_t<
+                      std::is_fundamental<T>::value || std::is_base_of<std::string, T>::value, int> = 0>
+    T const& at(Key<K, T> const& key) const {
+        static_assert(!std::is_const<T>::value,
+                      "Due to implementation constraints, const keys are not supported.");
+        try {
+            auto foo = unsafeLookup(key.getId());
+            return boost::get<T const&>(foo);
+        } catch (boost::bad_get const&) {
+            std::stringstream message;
+            message << "Key not found: " << key;
+            throw LSST_EXCEPT(pex::exceptions::OutOfRangeError, message.str());
```

Comment:
I'm a bit uncomfortable with this because it further erodes the abstraction that key-based queries ask "is there a T called X?". Do you need this information to be accessible to an exception handler? You could address the debugging use case just by giving different error messages.

Or, given our recent discussion of `operator==`, is the aforementioned definition of `get` and `contains` one that would make sense to C++ programmers, or have I picked an abstraction that makes sense only to me?

Reply:
I think either this behavior or the one I proposed would make sense to C++ programmers.  :+1: for a different error message, as I think the human-debugging use case is the important one.